### ISSUE TITLE
desired.properties: arrays are not supported in Simulator but on IoT Hub

### ISSUE THREAD

**Issue by derSchtefan** [Azure/iotedgehubdev#326]
In our system, we use a group of desired properties that are arrays with configuration items. It works very well in IoT hub, but when we try to test the system in the simulator it fails with this message:

`
WARNING: Fail to update XXMODULENAMEXX twin. Code:400. Detail:{"Message":"ErrorCode:ArgumentInvalid;Provided json is not a valid 'tags' or 'properties' object: arrays are not supported.","ExceptionMessage":"Tracking ID:487dbe89cfdd4c2a8f791389530c3361-G:29-TimeStamp:03/30/2021 14:16:03"}
`

Array are supported by the IoT Hub, as confirmed by running it there and also confirmed by this link:  https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-module-twins#tags-and-properties-format

The desired properties of the module in question look like this:
```
"XXMODULENAMEXX": {
    "properties.desired": {
        "SomeProperty": [
            { "Prop1":"ABC", "Prop2": "123" },
            { "Prop1":"ABC", "Prop2": "123" },
            { "Prop1":"ABC", "Prop2": "123" }
        ]
    }
},
```

A workaround would be to completely rewrite the configuration system, or to somehow fake it when running in the simulator, but that defeats the purpose of running it in the simulator, especially if it is part of our testing strategy.

Is there any plan to add array support ?

**Comment by smcd253** [comment]
Hi @derSchtefan,

Thanks for bringing this up. Support for arrays in desired properties sounds like an appropriate enhancement. We will put this in our backlog and will respond her when we address it.

**Comment by konichi3** [comment]
Hi @derSchtefan 

We would like to explore options for our potential solutions. Would you be able to share your scenario on using Array data type in desired properties? This would help us prioritize the issue accordingly.

thank you in advance.

**Comment by derSchtefan** [comment]
Whenever we use a list of items for a configuration element, that entry must be an array.
An example from our current project: We are reading measurements from multiple devices via the Microsoft OPC Publisher module, and process them in our edge module. We tell the edge module via its desired properties which elements belong into which category. So the config looks like

```json
"HotNodes" : [ 
  { "id": "xx", "category": "hot"} , 
  { "id": "yy", "category": "hot", "bla": "bla"} , 
  { "id": "xx", "category": "warm"} , 
  {"id": "*", "category": "other" }
],
"OtherSetting": { ... } 
```

One will use an array as value for a property always when one needs an ordered list of elements. I think this is quite straightforward and a common use case in information processing.

Some time ago, the IoT hub itself was not able to provide arrays as config parameters, but since quite a long time supports this now.

**Comment by konichi3** [comment]
@derSchtefan Thank you. We will investigate further and get back to you when we have an update.

thanks!

**Comment by smcd253** [comment]
Hi @derSchtefan, can you please post your deployment file and steps to reproduce this issue?

I have tried adding property arrays to a deployment file to both the edgeAgent module as well as an example module "filtermodule" and have had no issues.

Here is my deployment file:
```json
{
  "modulesContent": {
    "$edgeAgent": {
      "properties.desired": {
        "SomeProperty": [
          {
            "Prop1": "ABC",
            "Prop2": "123"
          },
          {
            "Prop1": "ABC",
            "Prop2": "123"
          },
          {
            "Prop1": "ABC",
            "Prop2": "123"
          }
        ],
        "schemaVersion": "1.0",
        "runtime": {
          "type": "docker",
          "settings": {
            "minDockerVersion": "v1.25",
            "loggingOptions": "",
            "registryCredentials": {}
          }
        },
        "systemModules": {
          "edgeAgent": {
            "type": "docker",
            "settings": {
              "image": "mcr.microsoft.com/azureiotedge-agent:1.0",
              "createOptions": "{}"
            }
          },
          "edgeHub": {
            "type": "docker",
            "status": "running",
            "restartPolicy": "always",
            "settings": {
              "image": "mcr.microsoft.com/azureiotedge-hub:1.0",
              "createOptions": "{\"HostConfig\":{\"PortBindings\":{\"5671/tcp\":[{\"HostPort\":\"5671\"}],\"8883/tcp\":[{\"HostPort\":\"8883\"}],\"443/tcp\":[{\"HostPort\":\"443\"}]}}}"
            }
          }
        },
        "modules": {
          "tempSensor": {
            "version": "1.0",
            "type": "docker",
            "status": "running",
            "restartPolicy": "always",
            "settings": {
              "image": "mcr.microsoft.com/azureiotedge-simulated-temperature-sensor:1.0",
              "createOptions": "{}"
            }
          },
          "filtermodule": {
            "properties.desired": {
              "SomeProperty": [
                {
                  "Prop1": "ABC",
                  "Prop2": "123"
                },
                {
                  "Prop1": "ABC",
                  "Prop2": "123"
                },
                {
                  "Prop1": "ABC",
                  "Prop2": "123"
                }
              ]
            },
            "version": "1.0",
            "type": "docker",
            "status": "running",
            "restartPolicy": "always",
            "settings": {
              "image": "localhost:5000/filtermodule:0.0.1-amd64",
              "createOptions": "{}"
            }
          }
        }
      }
    },
    "$edgeHub": {
      "properties.desired": {
        "schemaVersion": "1.0",
        "routes": {
          "sensorTofiltermodule": "FROM /messages/modules/tempSensor/outputs/temperatureOutput INTO BrokeredEndpoint(\"/modules/filtermodule/inputs/input1\")",
          "filtermoduleToIoTHub": "FROM /messages/modules/filtermodule/outputs/* INTO $upstream"
        },
        "storeAndForwardConfiguration": {
          "timeToLiveSecs": 7200
        }
      }
    }
  }
}
```

I have developed and run the simulator using the [pypi package](https://pypi.org/project/iotedgedev/) built from https://github.com/Azure/iotedgedev. Here are the commands I ran:

```bash
mkdir temp
cd temp
source <path_to_venv>/Scripts/activate
pip install iotedgedev
iotedgedev init
nano deployment.template.json
# insert desired properties into edgeAgent and filtermodule modules as is written by OP
# save (ctrl+x, ctrl+y)
iotedgedev build
iotedgedev simulator start
iotedgedev monitor
# quit (ctlr+c)
iotedgedev simulator stop
```
````

**Comment by derSchtefan** [comment]
Yes, @konichi3  and @smcd253, attached a file to reproduce it.
We are using VS Code with the latest extension (`vsciot-vscode.azure-iot-edge-1.24.0`) running the latest iotedgehubdev (0.14.6).

Looking at your example I can see that **you tested it with "properties.desired" at the wrong location**. The desired properites of each module are defined top level, not inside the desired properties of the edgeAgent (where the module is instantiated).


The message we are getting:
```
& "c:\Users\stefan\.vscode\extensions\vsciot-vscode.azure-iot-edge-1.24.0\iotedgehubdev\0.14.6\iotedgehubdev" start -d "c:\dev\symmedia\GF.MS.DB.Symmedia.Data.Modules\config\deployment.debug.amd64.json" -v
edgeHub image version is ignored in solution mode.
Network azure-iot-edge-dev is external, skipping
WARNING: Fail to update tempSense twin. Code:400. Detail:{"Message":"ErrorCode:ArgumentInvalid;Provided json is not a valid 'tags' or 'properties' object: arrays are not supported.","ExceptionMessage":"Tracking ID:120ce12b84eb469fb8eb31543d43d8f7-G:5-TimeStamp:05/01/2021 12:45:55"}
Pulling edgeHubDev ... done
Creating edgeHubDev ... done
Creating tempSense  ... done
Attaching to edgeHubDev, tempSense
edgeHubDev    | 2021-05-01 12:45:58  Starting Edge Hub
```

(See the WARNING message).

I can reproduce this with any module, with any config. For sake of client confidentiality I rebuilt the solution with the temperature sensor module:

```json
{
  "$schema-template": "2.0.0",
  "modulesContent": {
    "$edgeAgent": {
      "properties.desired": {
        "schemaVersion": "1.1",
        "runtime": {
          "type": "docker",
          "settings": {
            "minDockerVersion": "v1.25",
            "registryCredentials": {}
          }
        },
        "systemModules": {
          "edgeAgent": {
            "type": "docker",
            "settings": { "image": "mcr.microsoft.com/azureiotedge-agent:1.1", "createOptions": {} }
          },
          "edgeHub": {
            "type": "docker", "status": "running", "restartPolicy": "always",
            "settings": {
              "image": "mcr.microsoft.com/azureiotedge-hub:1.1", "createOptions": {} }
          }
        },
        "modules": {
          "tempSense": {
            "version": "1.0",
            "type": "docker",
            "status": "running",
            "restartPolicy": "always",
            "settings": {
              "image": "mcr.microsoft.com/azureiotedge-simulated-temperature-sensor:1.0",
              "createOptions": {}
            }
          }
        }
      }
    },
    "tempSense": {
      "properties.desired": {
        "HotNodes": [
          { "EndpointUrl": "*",             "NodeId": "nsmp:ABC:1"          },
          { "EndpointUrl": "*",             "NodeId": "nsmp:ABC:2"          },
          { "EndpointUrl": "*",             "NodeId": "nsmp:DEF:3"          },
          { "EndpointUrl": "*",             "NodeId": "nsmp:JKL:4"          },
          { "EndpointUrl": "opc.tcp://abc", "NodeId": "nsmp:XYZ:5"          },
          { "EndpointUrl": "opc.tcp://klm", "NodeId": "nsmp:XYZ:5"          }
        ]
      }
    },
    "$edgeHub": {
      "properties.desired": {
        "schemaVersion": "1.1",
        "routes": { },
        "storeAndForwardConfiguration": {          "timeToLiveSecs": 7200        }
      }
    }   
  }
}
```

**Comment by derSchtefan** [comment]
@konichi3 and @smcd253, any update? Have you seen my reply regarding the error with your testing?

**Comment by marianan** [comment]
@derSchtefan thank you for your reply. Apologies for the delay. We will investigate and respond back when we have updates. This item is currently in our backlog.

**Comment by marianan** [comment]
@derSchtefan the issue has been resolved in iotedgehubdev 0.14.11. Please try it out.
